polyinterface/polyinterface.py

```diff
@@ -53,36 +53,34 @@ def write(self, message):
     def flush(self):
         pass
 
+# Make a handler that writes to a file,
+# making a new file at midnight and keeping 3 backups
+_handler = logging.handlers.TimedRotatingFileHandler("./logs/debug.log", when="midnight", backupCount=30)
+def set_log_format(fmt_string='%(asctime)s %(threadName)-10s %(name)-18s %(levelname)-8s %(module)s:%(funcName)s: %(message)s'):
```

Comment:
As a general rule, I prefer to wrap string constants within some class, i.e. "PolyLogger" or some such and export class rather than function. But other than that, I have no issues with this update.

Reply:
Thanks, but it's in polyinterface._handler, right? (I'm still far from a python expert...)  I want to move all these to a different file because I don't like that setup_log is called every time you import polyinterface.

Reply:
I was thinking of something like this:

`import os
import logging
from logging import handlers as log_handlers
import warnings

class PolyLogger:

    LOGS_DIR= './logs'
    LOG_FILE = 'debug.log'
    ROTATION = 'midnight'
    LOGGER_NAME = 'polyinterface'
    WARN_LOGGER_NAME = 'py.warnings'
    BACKUP_COUNT = 30
    FMT_STRING = '%(asctime)s %(threadName)-10s %(name)-18s %(levelname)-8s %(module)s:%(funcName)s: %(message)s'

    def __init__(self):
        if not os.path.exists(PolyLogger.LOGS_DIR):
            os.makedirs(PolyLogger.LOGS_DIR)
        self.handler = log_handlers.TimedRotatingFileHandler(
                os.path.join(PolyLogger.LOGS_DIR, PolyLogger.LOG_FILE),
                when=PolyLogger.ROTATION, backupCount=PolyLogger.BACKUP_COUNT)
        logging.captureWarnings(True)
        self.set_log_format(PolyLogger.FMT_STRING)
        # Attach the handler to the logger
        logging.basicConfig(
            handlers=[self.handler],
            level=logging.DEBUG
        )

        self.logger = logging.getLogger(PolyLogger.LOGGER_NAME)
        self.logger.propagate = False # If True we get duplicates?
        self.logger.addHandler(self.handler)

        self.warnlog = logging.getLogger(PolyLogger.WARN_LOGGER_NAME)
        warnings.formatwarning = self.warning_on_one_line
        self.warnlog.addHandler(self.handler)

    def set_log_format(self, fmt_string):
        # Format each log message like this
        formatter = logging.Formatter(fmt_string)
        # Attach the formatter to the handler
        self.handler.setFormatter(formatter)

    @staticmethod
    def warning_on_one_line(message, category, filename, lineno, file=None, line=None):
        return '{}:{}: {}: {}'.format(filename, lineno, category.__name__, message)

LOG_HANDLER = PolyLogger()
LOGGER = LOG_HANDLER.logger`

Reply:
Yes, but then every time a "import polyinterface' is run it creates a new PolyLogger instance, right? which we don't want...